puffin/src/trace.rs

```diff
@@ -659,6 +702,106 @@ impl<PT: ProtocolTypes> fmt::Display for InputAction<PT> {
     }
 }
 
+/// This macro defines the precomputation syntax to add precomputations to an input action step
+///
+/// The following syntaxes are accepted :
+/// ```ignore
+/// # use puffin::input_action;
+/// # use puffin::term;
+/// # use puffin::trace::Precomputation;
+/// # use puffin::trace::InputAction;
+///
+/// input_action!{term!{fn_seq_0()}};
+/// input_action!{term!{fn_seq_0()} => term!{fn_seq_0()}};
+/// input_action!{"this_is_a_label" = term!{fn_seq_0()} => term!{fn_seq_0()}};
+/// input_action!{
+///     "this_is_a_label" = term!{fn_seq_0()} =>
+///         term!{fn_seq_0()} =>
+///             term!{fn_seq_0()}
+/// };
+/// // the latter is equivalent to
+/// input_action!{
+///     "this_is_a_label" = term!{fn_seq_0()}, term!{fn_seq_0()} =>
+///         term!{fn_seq_0()}
+/// };
+/// ```
+///
+/// All the previous examples respectively produce
+/// ```ignore
+/// # use puffin::trace::Precomputation;
+/// # use puffin::trace::InputAction;
+/// # use puffin::term;
+/// # use crate::algebra::test_signature::fn_seq_0;
+///
+/// InputAction {
+///     recipe: term!{fn_seq_0()},
+///     precomputations: vec![],
+/// };
+/// InputAction {
+///     recipe: term!{fn_seq_0()},
+///     precomputations: vec![Precomputation{label: "".into(), recipe: term!{fn_seq_0()}}],
+/// };
+/// InputAction {
+///     recipe: term!{fn_seq_0()},
+///     precomputations: vec![Precomputation{label: "this_is_a_label".into(), recipe: term!{fn_seq_0()}}],
+/// };
+/// InputAction {
+///     recipe: term!{fn_seq_0()},
+///     precomputations: vec![
+///         Precomputation{label: "this_is_a_label".into(), recipe: term!{fn_seq_0()}},
+///         Precomputation{label: "".into(), recipe: term!{fn_seq_0()}}
+///     ],
+/// };
+/// ```
```

Comment:
Maybe start the DOC with an actually useful example?

Reply:
I improved the documentation, I hope now it's more explicit